File: Osnovne skole/2026/tri_league_2026/solutions/main.cpp
#include <algorithm>
#include <iostream>
#include <map>
#include <string>
#include <vector>

using namespace std;

struct Mec {
    string timA;  // domacin
    string timB;
    string timC;
    int a = 0;
    int b = 0;
    int c = 0;
};

istream& operator>>(istream& ulaz, Mec& mec) {
    string crta1, crta2, dvotacka, crta3, crta4;
    ulaz >> mec.timA >> crta1 >> mec.timB >> crta2 >> mec.timC >> dvotacka >> mec.a >> crta3 >> mec.b >> crta4 >> mec.c;
    return ulaz;
}

ostream& operator<<(ostream& izlaz, const Mec& mec) {
    izlaz << mec.timA << " - " << mec.timB << " - " << mec.timC << " : " << mec.a << " - " << mec.b << " - " << mec.c;
    return izlaz;
}

struct StatistikaTima {
    string ime;
    int P = 0;  // liga bodovi
    int F = 0;  // postignuti fragovi
    int A = 0;  // primljeni fragovi
    int D = 0;  // F - A
    int W = 0;  // broj strogih prvih mjesta
    int H = 0;  // fragovi postignuti kao domacin
    int X = 0;  // ciste pobjede
    int brojDomacihUtakmica = 0;
};

static int uporediProsjekDomacihFragova(const StatistikaTima& lijevi, const StatistikaTima& desni) {
    // Ako tim nema nijednu domacu utakmicu, njegov prosjek domacih fragova je 0.
    if (lijevi.brojDomacihUtakmica == 0 && desni.brojDomacihUtakmica == 0) return 0;
    if (lijevi.brojDomacihUtakmica == 0) return -1;
    if (desni.brojDomacihUtakmica == 0) return 1;
    int lijevaStrana = lijevi.H * desni.brojDomacihUtakmica;
    int desnaStrana = desni.H * lijevi.brojDomacihUtakmica;
    if (lijevaStrana < desnaStrana) return -1;
    if (lijevaStrana > desnaStrana) return 1;
    return 0;
}

static bool istiKriterijiRangiranja(const StatistikaTima& lijevi, const StatistikaTima& desni) {
    if (lijevi.P != desni.P) return false;
    if (lijevi.D != desni.D) return false;
    if (lijevi.F != desni.F) return false;
    if (lijevi.W != desni.W) return false;
    if (lijevi.X != desni.X) return false;
    if ((lijevi.F - lijevi.H) != (desni.F - desni.H)) return false;
    return uporediProsjekDomacihFragova(lijevi, desni) == 0;
}

static bool boljiPoKriterijimaRangiranja(const StatistikaTima& lijevi, const StatistikaTima& desni) {
    if (lijevi.P != desni.P) return lijevi.P > desni.P;
    if (lijevi.D != desni.D) return lijevi.D > desni.D;
    if (lijevi.F != desni.F) return lijevi.F > desni.F;
    if (lijevi.W != desni.W) return lijevi.W > desni.W;
    if (lijevi.X != desni.X) return lijevi.X > desni.X;
    int gostujucaSnagaLijevog = lijevi.F - lijevi.H;
    int gostujucaSnagaDesnog = desni.F - desni.H;
    if (gostujucaSnagaLijevog != gostujucaSnagaDesnog) return gostujucaSnagaLijevog > gostujucaSnagaDesnog;
    int poredenjeProsjeka = uporediProsjekDomacihFragova(lijevi, desni);
    if (poredenjeProsjeka != 0) return poredenjeProsjeka < 0;
    return lijevi.ime < desni.ime;  // samo da poredak bude deterministicki
}

void primijeniMec(vector<StatistikaTima>& timovi, const map<string, int>& indeksiTimova, const Mec& mec) {
    int indeksA = indeksiTimova.at(mec.timA);
    int indeksB = indeksiTimova.at(mec.timB);
    int indeksC = indeksiTimova.at(mec.timC);

    int rezultati[3] = {mec.a, mec.b, mec.c};
    int indeksi[3] = {indeksA, indeksB, indeksC};

    timovi[indeksA].F += mec.a;
    timovi[indeksB].F += mec.b;
    timovi[indeksC].F += mec.c;

    timovi[indeksA].A += mec.b + mec.c;
    timovi[indeksB].A += mec.a + mec.c;
    timovi[indeksC].A += mec.a + mec.b;

    timovi[indeksA].D = timovi[indeksA].F - timovi[indeksA].A;
    timovi[indeksB].D = timovi[indeksB].F - timovi[indeksB].A;
    timovi[indeksC].D = timovi[indeksC].F - timovi[indeksC].A;

    timovi[indeksA].H += mec.a;
    timovi[indeksA].brojDomacihUtakmica++;

    vector<pair<int, int>> poredak = {{rezultati[0], 0}, {rezultati[1], 1}, {rezultati[2], 2}};
    sort(poredak.begin(), poredak.end(), greater<pair<int, int>>());

    const int maksimum = poredak[0].first;
    const int srednji = poredak[1].first;
    const int minimum = poredak[2].first;

    if (maksimum == srednji && srednji == minimum) {
        timovi[indeksi[0]].P += 2;
        timovi[indeksi[1]].P += 2;
        timovi[indeksi[2]].P += 2;
    } else if (maksimum == srednji) {
        timovi[indeksi[poredak[0].second]].P += 3;
        timovi[indeksi[poredak[1].second]].P += 3;
    } else if (srednji == minimum) {
        timovi[indeksi[poredak[0].second]].P += 4;
        timovi[indeksi[poredak[1].second]].P += 1;
        timovi[indeksi[poredak[2].second]].P += 1;
    } else {
        timovi[indeksi[poredak[0].second]].P += 4;
        timovi[indeksi[poredak[1].second]].P += 2;
    }

    if (maksimum > srednji) {
        int lokalniPobjednik = poredak[0].second;
        StatistikaTima& pobjednik = timovi[indeksi[lokalniPobjednik]];
        pobjednik.W++;
        if (maksimum >= 2 * minimum) pobjednik.X++;
    }
}

vector<StatistikaTima> izracunajPoredak(const vector<string>& imenaTimova, const vector<Mec>& mecevi) {
    map<string, int> indeksiTimova;
    vector<StatistikaTima> timovi;
    timovi.reserve(imenaTimova.size());
    for (int i = 0; i < static_cast<int>(imenaTimova.size()); i++) {
        indeksiTimova[imenaTimova[i]] = i;
        StatistikaTima statistika;
        statistika.ime = imenaTimova[i];
        timovi.push_back(statistika);
    }
    for (const Mec& mec : mecevi) primijeniMec(timovi, indeksiTimova, mec);
    sort(timovi.begin(), timovi.end(), boljiPoKriterijimaRangiranja);
    return timovi;
}

bool trebaRazigravanje(const vector<StatistikaTima>& poredaniTimovi) {
    for (int i = 1; i < static_cast<int>(poredaniTimovi.size()); i++) {
        if (istiKriterijiRangiranja(poredaniTimovi[i - 1], poredaniTimovi[i])) return true;
    }
    return false;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int brojTimova = 0;
    cin >> brojTimova;
    vector<string> imenaTimova(brojTimova);
    for (int i = 0; i < brojTimova; i++) cin >> imenaTimova[i];

    int brojUtakmica = 0;
    cin >> brojUtakmica;
    vector<Mec> mecevi(brojUtakmica);
    for (int i = 0; i < brojUtakmica; i++) cin >> mecevi[i];

    vector<StatistikaTima> poredak = izracunajPoredak(imenaTimova, mecevi);
    if (trebaRazigravanje(poredak)) {
        cout << "Potrebno je razigravanje!\n";
    } else {
        for (const StatistikaTima& tim : poredak) {
            cout << tim.ime << " " << tim.P << "\n";
        }
    }
    return 0;
}

File: Osnovne skole/2026/tri_league_2026/solutions/main.py
import sys
from functools import cmp_to_key

input = sys.stdin.readline


def parse_mec(line):
    # Format: timA - timB - timC : a - b - c
    parts = line.split()
    # parts: [timA, '-', timB, '-', timC, ':', a, '-', b, '-', c]
    timA = parts[0]
    timB = parts[2]
    timC = parts[4]
    a = int(parts[6])
    b = int(parts[8])
    c = int(parts[10])
    return timA, timB, timC, a, b, c


def uporedi_prosjek_domacih_fragova(lijevi, desni):
    if lijevi['brojDomacihUtakmica'] == 0 and desni['brojDomacihUtakmica'] == 0:
        return 0
    if lijevi['brojDomacihUtakmica'] == 0:
        return -1
    if desni['brojDomacihUtakmica'] == 0:
        return 1
    lijeva_strana = lijevi['H'] * desni['brojDomacihUtakmica']
    desna_strana = desni['H'] * lijevi['brojDomacihUtakmica']
    if lijeva_strana < desna_strana:
        return -1
    if lijeva_strana > desna_strana:
        return 1
    return 0


def isti_kriteriji_rangiranja(lijevi, desni):
    if lijevi['P'] != desni['P']:
        return False
    if lijevi['D'] != desni['D']:
        return False
    if lijevi['F'] != desni['F']:
        return False
    if lijevi['W'] != desni['W']:
        return False
    if lijevi['X'] != desni['X']:
        return False
    if (lijevi['F'] - lijevi['H']) != (desni['F'] - desni['H']):
        return False
    return uporedi_prosjek_domacih_fragova(lijevi, desni) == 0


def bolji_po_kriterijima(lijevi, desni):
    if lijevi['P'] != desni['P']:
        return -1 if lijevi['P'] > desni['P'] else 1
    if lijevi['D'] != desni['D']:
        return -1 if lijevi['D'] > desni['D'] else 1
    if lijevi['F'] != desni['F']:
        return -1 if lijevi['F'] > desni['F'] else 1
    if lijevi['W'] != desni['W']:
        return -1 if lijevi['W'] > desni['W'] else 1
    if lijevi['X'] != desni['X']:
        return -1 if lijevi['X'] > desni['X'] else 1
    gost_lijevi = lijevi['F'] - lijevi['H']
    gost_desni = desni['F'] - desni['H']
    if gost_lijevi != gost_desni:
        return -1 if gost_lijevi > gost_desni else 1
    cmp = uporedi_prosjek_domacih_fragova(lijevi, desni)
    if cmp != 0:
        return 1 if cmp < 0 else -1
    if lijevi['ime'] < desni['ime']:
        return -1
    if lijevi['ime'] > desni['ime']:
        return 1
    return 0


def novi_tim(ime):
    return {'ime': ime, 'P': 0, 'F': 0, 'A': 0, 'D': 0, 'W': 0, 'H': 0, 'X': 0, 'brojDomacihUtakmica': 0}


def primijeni_mec(timovi, indeksi_timova, timA, timB, timC, a, b, c):
    iA = indeksi_timova[timA]
    iB = indeksi_timova[timB]
    iC = indeksi_timova[timC]

    timovi[iA]['F'] += a
    timovi[iB]['F'] += b
    timovi[iC]['F'] += c

    timovi[iA]['A'] += b + c
    timovi[iB]['A'] += a + c
    timovi[iC]['A'] += a + b

    timovi[iA]['D'] = timovi[iA]['F'] - timovi[iA]['A']
    timovi[iB]['D'] = timovi[iB]['F'] - timovi[iB]['A']
    timovi[iC]['D'] = timovi[iC]['F'] - timovi[iC]['A']

    timovi[iA]['H'] += a
    timovi[iA]['brojDomacihUtakmica'] += 1

    rezultati = [(a, 0), (b, 1), (c, 2)]
    rezultati.sort(key=lambda x: -x[0])
    maksimum = rezultati[0][0]
    srednji = rezultati[1][0]
    minimum = rezultati[2][0]
    idx = [iA, iB, iC]

    if maksimum == srednji == minimum:
        timovi[idx[0]]['P'] += 2
        timovi[idx[1]]['P'] += 2
        timovi[idx[2]]['P'] += 2
    elif maksimum == srednji:
        timovi[idx[rezultati[0][1]]]['P'] += 3
        timovi[idx[rezultati[1][1]]]['P'] += 3
    elif srednji == minimum:
        timovi[idx[rezultati[0][1]]]['P'] += 4
        timovi[idx[rezultati[1][1]]]['P'] += 1
        timovi[idx[rezultati[2][1]]]['P'] += 1
    else:
        timovi[idx[rezultati[0][1]]]['P'] += 4
        timovi[idx[rezultati[1][1]]]['P'] += 2

    if maksimum > srednji:
        lokalni_pobjednik = rezultati[0][1]
        pobjednik = timovi[idx[lokalni_pobjednik]]
        pobjednik['W'] += 1
        if maksimum >= 2 * minimum:
            pobjednik['X'] += 1


def main():
    broj_timova = int(input())
    imena_timova = [input().strip() for _ in range(broj_timova)]

    broj_utakmica = int(input())
    mecevi = []
    for _ in range(broj_utakmica):
        mecevi.append(parse_mec(input()))

    indeksi_timova = {ime: i for i, ime in enumerate(imena_timova)}
    timovi = [novi_tim(ime) for ime in imena_timova]

    for mec in mecevi:
        primijeni_mec(timovi, indeksi_timova, *mec)

    timovi.sort(key=cmp_to_key(bolji_po_kriterijima))

    treba_razigravanje = False
    for i in range(1, len(timovi)):
        if isti_kriteriji_rangiranja(timovi[i - 1], timovi[i]):
            treba_razigravanje = True
            break

    if treba_razigravanje:
        print("Potrebno je razigravanje!")
    else:
        for tim in timovi:
            print(tim['ime'], tim['P'])


main()
